samples/allocator/enclave/enc.cpp

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <queue>
+
+// Include the trusted allocator_demo header that is generated
+// during the build. This file is generated by calling the
+// sdk tool oeedger8r against the allocator_demo.edl file.
+#include "allocator_demo_t.h"
+
+static uint64_t _max_allocation_size = 2 * 1024;
+
+void enclave_configure(uint32_t seed, uint64_t max_allocation_size)
+{
+    // Initialize random number generator.
+    srand(seed);
+    _max_allocation_size = max_allocation_size;
+}
+
+#define QUEUE_LENGTH 15
+
+void enclave_thread(uint64_t num_iterations)
+{
+    std::queue<void*> allocations;
```

Comment:
@mingweishih This is the function I'm using to demonstrate performance improvements with pluggable allocators. Do you have suggestions for something else (or additionally) we could do to here, *without* making the sample too complicated?

Reply:
@achamayou 

I get at-best 2X speed up.

```
host/allocator_demo_host enclave/enclave_default.signed enclave/enclave_custom.signed --num-threads 16 --max-allocation-size 64
Configuration:
    simulate = false
    num-threads = 16
    num-iterations = 1000000
    seed = 1594166907
    max-allocation-size = 64


Running enclave linked with default allocator
Loading enclave enclave/enclave_default.signed
Elapsed time = 18629 milliseconds
Unloading enclave

Running enclave linked with oesnmalloc (pluggable allocator)
Loading enclave enclave/enclave_custom.signed
Elapsed time = 9555 milliseconds
Unloading enclave
```

Is there a some allocation pattern that would show more speedup?

Reply:
This implementation looks good to me. However, I feel this sample is more like a benchmark other than a program that a user would directly consume/modify. If the purpose is the former, then I think it's totally fine.

Reply:
@anakrish how many cores does your machine have? Something snmalloc is particularly good at is allocating in one thread and deallocating in another (common in message passing-style code).

Reply:
@achamayou Seems to have been something wrong on my machine. After reboot and clean up of the sample, I now get the following output which shows anywhere from 2 to 40x speed up depending upon number of threads.

```
Configuration:
                                mode = hardware
        num-allocations (per-thread) = 100000
                 max-allocation-size = 16384 bytes
num-threads = 1:
      dlmalloc   (default allocator) =    9 milliseconds
    oesnmalloc (pluggable allocator) =    4 milliseconds

num-threads = 3:
      dlmalloc   (default allocator) =   62 milliseconds
    oesnmalloc (pluggable allocator) =    6 milliseconds

num-threads = 5:
      dlmalloc   (default allocator) =  179 milliseconds
    oesnmalloc (pluggable allocator) =   12 milliseconds

num-threads = 7:
      dlmalloc   (default allocator) =  339 milliseconds
    oesnmalloc (pluggable allocator) =   13 milliseconds

num-threads = 9:
      dlmalloc   (default allocator) =  531 milliseconds
    oesnmalloc (pluggable allocator) =   17 milliseconds

num-threads = 11:
      dlmalloc   (default allocator) =  784 milliseconds
    oesnmalloc (pluggable allocator) =   21 milliseconds

num-threads = 13:
      dlmalloc   (default allocator) = 1042 milliseconds
    oesnmalloc (pluggable allocator) =   24 milliseconds

num-threads = 15:
      dlmalloc   (default allocator) = 1249 milliseconds
    oesnmalloc (pluggable allocator) =   30 milliseconds


```

Reply:
@anakrish this is a much greater difference than we see in CCF because this is extremely allocation heavy of course, and CCF spends less of its time on allocation proportionally-speaking obviously. What it does illustrate well however, is how badly dlmalloc fares when scaling up threads:cores, to the point where there's barely any throughput benefit.

Again, this is application dependent, but in our case (https://github.com/openenclave/openenclave/pull/2649), scaling past two execution threads was impossible as a result.